fix ranking rules after _geo do not work

HDT3213 opened the pull request:
# Pull Request

## Related issue
Fixes #5405

## What does this PR do?

`GeoSort::next_bucket` now returns all points with the same distance to the searched point at a time, so that ranking rules after _geo can do their work 

I built a test dataset that was larger than GeoSort’s cache size and used it to verify that my code could refill the cache after it was exhausted.

Since the  ranking rules after geo may select any doc in the last bucket, `next_bucket` must always return a complete bucket to ensure the correct result, even if this makes the total amount exceed `limit+offset`.

Fortunately, `bucket_sort` will stop iterating `next_bucket` when the total returned exceeds `limit+offset`, so we don't need to worry about "unbounded" work

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
+                                score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                    target_point: self.point,
+                                    ascending: self.ascending,
+                                    value: Some(point0.to_owned()),
```
Outputting the coordinates of one random points in the bucket is pretty bad and could be misleading.
We should stop returning this value entirely and only output the distance of this bucket from the required point.

But that would be a breaking change and cannot be done in this PR. I’ll open an issue for v2.

HDT3213 replied:
OK, I'll leave this question alone for now.

By the way, as this [comment](https://github.com/meilisearch/meilisearch/issues/5405#issuecomment-2774762727) says, small differences in distance should be treated as equal. Perhaps we should allow users to customize this difference

irevoire replied:
We should think about it, but it’s more of a product question.

I think your PR is great as meilisearch returns the `_geoDistance` in meter it makes sense that when two documents share the same `_geoDistance` they’re bucketed together 👍

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
+                            }),
+                        }));
+                    } else {
+                        return Ok(Some(RankingRuleOutput {
+                            query,
+                            candidates: universe.clone(),
+                            score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                target_point: self.point,
+                                ascending: self.ascending,
+                                value: None,
+                            }),
+                        }));
```
Hey, I don’t understand why we’re returning nothing here?
Shouldn’t we continue with the loop and check if we have other candidates matching our bucket?

It would be nice if you could add a test where two candidates must be grouped together but are not part of the same bucket, as this is now a bit tricky to handle.

HDT3213 replied:
If self.cached_sorted_docids is still empty after refill, it means we have exhausted the entire candidate set. At this time, if current_bucket is not empty, I should return current_bucket. But if current_bucket is empty, I can only return an empty result set.

Also, when you say "grouped together", do you mean they appear in the cache at the same time?

irevoire replied:
Aaaah yes sorry my bad, I thought the else was applied to the other branch my bad 🤦 
So you’re actually doing what I was expecting 👌

irevoire reviewed (changes requested):
Hey, thanks for your contribution!

Could you add a test with two points pointing to the same spot in `crates/milli/src/search/new/tests/geo_sort.rs`.
By looking at the other test you’ll also see that you can select the scoring strategy to use the rtree or not so you should run your test twice with and without the rtree!

HDT3213 commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
+        let mut current_bucket = RoaringBitmap::new();
+        // current_distance stores the first point and distance in current bucket
+        // The farthest distance between two points on earth is about 2e7 meters, u32 is big enough to hold any distance.
+        let mut current_distance: Option<([f64; 2], u32)> = None;
```
In fact, distance can be represented by float. I originally wanted to use distance as the key of a map and then use u32. Now using u32 is a mistake.

irevoire replied:
Yes why not.
But when comparing floats do not forget that it’s recommended to run `left - right < f64::EPSILON` instead of a normal `==`.

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
+                                    value: Some(point0.to_owned()),
+                                }),
+                            }));
+                        } else {
+                            // same distance, point belongs to current bucket
+                            current_bucket.push(id);
+                            // remove from cadidates to prevent it from being added to the cache again
+                            geo_candidates.remove(id);
+                        }
```
While chatting with the team, we thought about this else and the way it could loop through the whole dataset if all documents share the same distance.
As a security hatch, we think it would be great if you could exit the loop anyway once we reach a certain number of documents in the same bucket.

1000 seems like a good value, but it would be awesome if you could make the thresholds configurable so we can easily test it out

HDT3213 replied:
I am worried that this will cause another problem. The ranking rules after geo need to get all the documents to sort correctly.If we only give 1000 documents, it may lead to wrong results.

For example, let's assume that all documents have the same distance, and the user sorts ["geoPoint...", "score:desc"]. Due to the security hatch, the document with the highest score may not be able to enter the bucket. in this case, the first item in the sorted result is not the document with the highest score.

irevoire replied:
Yes, the idea is that;
- Generally speaking there is no reason to have more than a few documents exactly at the same distance, that’s really an edge case
- If a malevolent actor notices want to DOS an instance, they could push a ton of document exactly in the same spot and then every search requests would be irresponsive 

So, we think it’s better to get results slightly wrong in a case that should never happen that risk a DOS 😅

HDT3213 commented:
I have added test cases as requested, and added max_bucket_size and distance_error_margin to GeoSort to limit the number of points in a bucket and the maximum distance (they default to 1000 and 1 meter). 

And I found that I had mistakenly used the RoaringBitmap::push instead of the RoaringBitmap::insert. This error was not caught by the previous tests, which could be dangerous.

HDT3213 commented:
To be honest, I don't know what my code has to do with the proxy😮‍💨

```
failures:

---- search::multi::proxy::error_remote_does_not_answer stdout ----
{
  "remotes": {
    "ms0": {
      "url": "http://127.0.0.1:46263/"
    },
    "ms1": {
      "url": "http://127.0.0.1:44643/"
    },
    "ms2": {
      "url": "https://thiswebsitedoesnotexist.example/"
    }
  }
}
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: error_remote_does_not_answer
Source: crates/meilisearch/tests/search/multi/proxy.rs:1327
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
   35    35 │   "offset": 0,
   36    36 │   "estimatedTotalHits": 3,
   37    37 │   "remoteErrors": {
   38    38 │     "ms2": {
   39       │-      "message": "error sending request",
   40       │-      "code": "remote_could_not_send_request",
         39 │+      "message": "remote host did not answer before the deadline",
         40 │+      "code": "remote_timeout",
   41    41 │       "type": "system",
   42       │-      "link": "https://docs.meilisearch.com/errors#remote_could_not_send_request"
         42 │+      "link": "https://docs.meilisearch.com/errors#remote_timeout"
   43    43 │     }
   44    44 │   }
   45    45 │ }
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.

thread 'search::multi::proxy::error_remote_does_not_answer' panicked at /github/home/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.39.0/src/runtime.rs:562:9:
snapshot assertion for 'error_remote_does_not_answer' failed in line 1327
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: insta::runtime::assert_snapshot
   3: integration::search::multi::proxy::error_remote_does_not_answer::{{closure}}
   4: tokio::task::local::LocalSet::run_until::{{closure}}
   5: core::ops::function::FnOnce::call_once
```

HDT3213 commented:
Hi @irevoire , I have completed the changes but I am getting some unrelated test failures. Actually I am also getting some random test failures in the main branch. So how should we handle these test failures?

irevoire commented on crates/milli/src/search/new/tests/geo_sort.rs:
```diff
@@ -95,6 +95,47 @@ fn test_geo_sort() {
+        AscDesc::Asc(Member::Geo([0., 0.])),
+        AscDesc::Desc(Member::Field("score".to_string())),
+    ]);
+    let (ids, scores) = execute_iterative_and_rtree_returns_the_same(&rtxn, &index, &mut s);
+    insta::assert_snapshot!(format!("{ids:?}"), @"[6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 1, 4, 3, 2, 5]");
+    insta::assert_snapshot!(format!("{scores:#?}"));
+
+    s.sort_criteria(vec![
+        AscDesc::Desc(Member::Geo([0., 0.])),
+        AscDesc::Desc(Member::Field("score".to_string())),
+    ]);
+    let (ids, scores) = execute_iterative_and_rtree_returns_the_same(&rtxn, &index, &mut s);
+    insta::assert_snapshot!(format!("{ids:?}"), @"[12, 13, 14, 15, 6, 7, 8, 9, 10, 11, 1, 4, 3, 2, 5]");
+    insta::assert_snapshot!(format!("{scores:#?}"));
+}
+
```
These tests are awesome.
I was just expecting two more test with a very small bucket of like 2 elements.
Let me know if it makes sense and if you can add them?

Then I think we’ll be good to merge 👌

HDT3213 replied:
In fact, execute_iterative_and_rtree_returns_the_same covers both the cases where the bucket size is greater than the cache size (need to refill the cache) and less than the cache size (no refill is required). I wonder which case a very small bucket is for?

irevoire replied:
Huum from what I see we have a test where the cache is small and must be refilled before we reach our thresholds. That’s good.
But I don’t see any test doing the opposite, basically, we have a very low threshold and must stop early even though the bucket is not empty

HDT3213 replied:
Hi, @irevoire I hope you had a great holiday.

You wanted to test the case where max_bucket_size is reached, right? I think I misunderstood you earlier, I will add the corresponding tests later.

irevoire replied:
Yes exactly!
Thank you 🙏 
Since there is two caches at the same time it’s not easy to explain ahah 😅

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -73,6 +71,26 @@ impl Strategy {
+    1000
+}
+
+#[cfg(test)]
+static DEFAULT_MAX_BUCKET_SIZE: std::sync::Mutex<u64> = std::sync::Mutex::new(1000);
+
+#[cfg(test)]
+pub fn set_default_max_bucket_size(n: u64) {
+    let mut size = DEFAULT_MAX_BUCKET_SIZE.lock().unwrap();
+    *size = n;
+}
+
+#[cfg(test)]
+fn default_max_bucket_size() -> u64 {
+    let max_size = *(DEFAULT_MAX_BUCKET_SIZE.lock().unwrap());
+    max_size
+}
```
Hey, sorry but since multiple tests are run in parallel sharing the same variable this will be super flaky.
One test may change the default max bucket size for the other 😭 

Maybe the easiest thing is to just shove the value into the `Strategy` like we did for the cache size.
Either you just add it to every variant, or you could create a structure `Parameter` that contains both the max_bucket_size and the cache_size + a new unit enum `Strategy` that doesn’t hold any data.

When calling this value from Meilisearch we always set it to `default` iirc, so you can do whatever you prefer it’s only local to this file.

HDT3213 replied:
😂 I think I would prefer to add a Parameter to hold all the relevant parameters

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -41,6 +39,22 @@ fn facet_number_values<'a>(
     Ok(iter)
 }
 
+#[derive(Debug, Clone, Copy)]
+
+pub struct Parameter {
```
```suggestion
#[derive(Debug, Clone, Copy)]
pub struct Parameter {
```

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -84,24 +98,32 @@ pub struct GeoSort<Q: RankingRuleQueryTrait> {
 impl<Q: RankingRuleQueryTrait> GeoSort<Q> {
     pub fn new(
-        strategy: Strategy,
+        parameter: &Parameter,
```
```suggestion
        parameter: Parameter,
```

When a structure derive `Copy` it’s generally best to take it by copy instead of reference

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -84,24 +98,32 @@ pub struct GeoSort<Q: RankingRuleQueryTrait> {
         Ok(Self {
             query: None,
-            strategy,
+            strategy: *strategy,
```
```suggestion
            strategy,
```

irevoire commented on crates/milli/src/search/new/geo_sort.rs:
```diff
@@ -84,24 +98,32 @@ pub struct GeoSort<Q: RankingRuleQueryTrait> {
             field_ids: None,
             rtree: None,
             cached_sorted_docids: VecDeque::new(),
+            max_bucket_size: *max_bucket_size,
+            distance_error_margin: *distance_error_margin,
```
```suggestion
            max_bucket_size,
            distance_error_margin,
```

irevoire commented on crates/milli/src/search/new/mod.rs:
```diff
@@ -299,7 +300,7 @@ fn get_ranking_rules_for_placeholder_search<'ctx>(
                     &mut sorted_fields,
                     &mut geo_sorted,
-                    geo_strategy,
+                    &geo_param,
```
```suggestion
                    geo_param,
```

irevoire commented on crates/milli/src/search/new/mod.rs:
```diff
@@ -375,7 +376,7 @@ fn get_ranking_rules_for_vector<'ctx>(
                     &mut sorted_fields,
                     &mut geo_sorted,
-                    geo_strategy,
+                    &geo_param,
```
```suggestion
                    geo_param,
```

irevoire commented on crates/milli/src/search/new/mod.rs:
```diff
@@ -477,7 +478,7 @@ fn get_ranking_rules_for_query_graph_search<'ctx>(
                     &mut sorted_fields,
                     &mut geo_sorted,
-                    geo_strategy,
+                    &geo_param,
```
```suggestion
                    geo_param,
```

irevoire reviewed (changes requested):
Awesome, one little comment, a type implementing `Copy` is generally not shared by ref but by copy directly
